File: Linked Lists/206. Reverse Linked List/Question_22.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def reverseList(self, head):
        current = head
        prev = None
        n = None
        while current:
            n = current.next
            current.next = prev
            prev = current
            current = n
        return prev 
    
def insert(head, val):
    if not head:
        return ListNode(val)
    current = head
    while current.next:
        current = current.next
    current.next = ListNode(val)
    return head

def printList(head):
    current = head
    while current:
        print(f"{current.val} -> ", end="")
        current = current.next
    print("NULL")

if __name__ == "__main__":
    head = None
    head = insert(head, 1)
    head = insert(head, 2)
    head = insert(head, 3)
    head = insert(head, 4)
    head = insert(head, 5)
    head = insert(head, 6)

    printList(head)

    solution = Solution()
    head = solution. reverseList(head)

    printList(head)

File: Linked Lists/206. Reverse Linked List/Question_22.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode* prev = NULL; 
    struct ListNode* current = head; 
    struct ListNode* next = NULL; 

    while (current != NULL) {
        next = current->next;
        current->next = prev; 
        prev = current;       
        current = next;      
    }

    return prev;
}
struct ListNode* recursiveReverseList(struct ListNode* head){
    if (head == NULL || head->next == NULL) {
        return head;
    }
    else{
        struct ListNode* newHead = recursiveReverseList(head->next); 
        head->next->next = head;
        head->next = NULL;       

        return newHead;  
    }

}
void insert( struct ListNode** head, int val){
    struct ListNode* temp = (struct ListNode*)malloc(sizeof(struct ListNode));
    (*temp).val = val;
    (*temp).next = NULL;

    if(*head == NULL)
    {
        *head  = temp;
    }
    else{
        struct ListNode* current = *head;
        while ((*current).next != NULL){
            current = (*current).next;
            
        }
        (*current).next = temp;
    }
}
void printList(struct ListNode* head) {
    struct ListNode* current = head;
    while (current != NULL) {
        printf("%d -> ", current->val);
        current = current->next;
    }
    printf("NULL\n");
}

void freeList(struct ListNode* head) {
    struct ListNode* current = head;
    while (current != NULL) {
        struct ListNode* next = current->next;
        free(current);
        current = next;
    }
}

int main() {
    struct ListNode* head = NULL;

    insert(&head, 1);
    insert(&head, 2);
    insert(&head, 3);
    insert(&head, 4);
    insert(&head, 5);
    insert(&head, 6);

    printf("Original list:\n");
    printList(head);

    head = recursiveReverseList(head);
    printf("Recursively reversed list:\n");
    printList(head);


    head = reverseList(head);
    printf("Iteratively reversed list:\n");
    printList(head);


    freeList(head);

    return 0;
}
